lib/services/android-debug-service.ts

```diff
@@ -43,14 +39,60 @@ class AndroidDebugService implements IDebugService {
 			: this.debugOnDevice();
 	}
 
-	public debugOnEmulator(): IFuture<void> {
+	private debugOnEmulator(): IFuture<void> {
 		return (() => {
 			this.$platformService.deployOnEmulator(this.platform).wait();
 			this.debugOnDevice().wait();
 		}).future<void>()();
 	}
 
-	public debugOnDevice(): IFuture<void> {
+	private isPortAvailable(candidatePort: number): IFuture<boolean> {
+		let future = new Future<boolean>();
+		let server = net.createServer();
+		server.on("error", (err: Error) => {
+			future.return(false);
+		});
+		server.listen(candidatePort, (err: Error) => {
+			server.once("close", () => {
+				future.return(true);
+			});
+			server.close();
+		});
+
+		return future;
+	}
+
+	private getForwardedLocalDebugPortForPackageName(deviceId: string, packageName: string): IFuture<number> {
+		return (() => {
+			let port = -1;
+			let forwardsResult = this.device.adb.executeCommand(["forward", "--list"]).wait();
+
+			//matches 123a188909e6czzc tcp:40001 localabstract:org.nativescript.testUnixSockets-debug
+			let regexp = new RegExp(`(${deviceId} tcp:)([\\d])+(?= localabstract:${packageName}-debug)`, "g");
```

Comment:
good one.
the best way should have been with a negative lookbehind... Not sure if exec is the optimal way thus I have done this with substring, which is ugly.

I leave it to you guys to decide.